src/payment/strategies/masterpass/masterpass-payment-strategy.ts

```diff
@@ -0,0 +1,127 @@
+        private _masterpassScriptLoader: MasterpassScriptLoader
+    ) {
+        super(store);
+    }
+
+    initialize(options: PaymentInitializeOptions): Promise<InternalCheckoutSelectors> {
+        const { methodId, masterpass: masterpassOptions } = options;
+
+        if (!masterpassOptions) {
+            throw new InvalidArgumentError('Unable to initialize payment because masterpass options is missing');
+        }
+
+        const state = this._store.getState();
+        this._paymentMethod = state.paymentMethods.getPaymentMethod(methodId);
+
+        return Promise.resolve(this._hasPaymentInfo()).then(hasPaymentInfo => {
+            if (!hasPaymentInfo) {
+                return this._masterpassClientSetup().then(checkoutCallback => checkoutCallback());
+            }
+
+            this._paymentGateway = masterpassOptions.gateway;
+
+            if (masterpassOptions.onPaymentSelect) {
+                masterpassOptions.onPaymentSelect();
+            }
+        }).then(() => {
+            return super.initialize(options);
+        });
+    }
+
+    deinitialize(options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        this._paymentMethod = undefined;
+        this._paymentGateway = undefined;
+
+        return super.deinitialize(options);
+    }
+
+    execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        const { payment, ...order } = payload;
+
+        if (!(payment && payment.paymentData)) {
+            throw new PaymentArgumentInvalidError(['payment.paymentData']);
+        }
+
+        const { paymentData } = payment;
+        const methodId = this._paymentGateway;
+
+        if (!methodId) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        return this._store.dispatch(this._orderActionCreator.submitOrder(order, options))
+            .then(() => this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(methodId)))
+            .then(() => this._store.dispatch(this._paymentActionCreator.submitPayment({ methodId, paymentData })));
+    }
+
+    private _createMasterpassPayload(paymentMethod: PaymentMethod, checkout: Checkout, storeConfig: StoreConfig): MasterpassCheckoutOptions {
+        return {
+            checkoutId: paymentMethod.initializationData.checkoutId,
+            allowedCardTypes: paymentMethod.initializationData.allowedCardTypes,
+            amount: checkout.subtotal.toFixed(2),
+            currency: storeConfig.currency.code,
+            cartId: checkout.cart.id,
+            suppressShippingAddress: true,
+        };
+    }
+
+    private _hasPaymentInfo(): boolean {
+        return this._paymentMethod
+            && this._paymentMethod.initializationData
+            && this._paymentMethod.initializationData.paymentData
+            && this._paymentMethod.initializationData.paymentData.nonce;
```

Comment:
Why are you introducing a new field called `nonce` under `initializationData`? If you look at our API contract, you'll see that under `PaymentMethod` already has a `nonce` field. You shouldn't need to introduce a new one.

Reply:
The only reason why we are doing this is that we are following the same pattern used on Braintree Visa Checkout https://github.com/bigcommerce/checkout-sdk-js/blob/fd43113e2c4c91c3f630949b3dcdb7b860cac73f/src/payment/strategies/braintree/braintree-visacheckout-payment-strategy.ts#L103 and that way is how the stripe payment provider expects to receive the payment data.

The only provider which uses the nonce at the root level is braintree but it has a custom code which is not the same approach that we are trying to follow for the masterpass implementation.

Reply:
I see. In fact, Braintree is the correct way to expose `nonce`. All other methods (Stripe, Visa Checkout) replicate the wrong approach, which unfortunately becomes the new norm. Anyway, considering it is not something new that you've introduced, I guess we can address this later.

Reply:
Actually, using the `initializationData` is what's really custom. It was made as a way of keeping extra data for the payment method. The `nonce` is part of the normal `paymentProvider` payload and it should be contained there.

VisaCheckout is an old implementation, not attached to any standard. We shouldn't consider it a model.

Reply:
I guess `initializationData` is undocumented, we could probably leave it as it is considering the deadline they're facing. We could leave a comment here to say `TODO: Refactor the API endpoint to return nonce in the right place.`